tests/test_models.py

```diff
@@ -3055,6 +3054,183 @@ def test_permission_override_on_readded_contributor(self):
+
+    def test_registration_children_have_correct_root(self):
+        registration_child = NodeFactory(parent=self.registration)
+        assert_equal(registration_child.root._id, self.registration._id)
+
+    def test_registration_grandchildren_have_correct_root(self):
+        registration_child = NodeFactory(parent=self.registration)
+        registration_grandchild = NodeFactory(parent=registration_child)
+
+        assert_equal(registration_grandchild.root._id, self.registration._id)
+
+    def test_fork_has_own_root(self):
+        fork = self.project.fork_node(auth=self.auth)
+        assert_equal(fork.root._id, fork._id)
+
+    def test_fork_children_have_correct_root(self):
+        fork = self.project.fork_node(auth=self.auth)
+        fork_child = NodeFactory(parent=fork)
+        assert_equal(fork_child.root._id, fork._id)
+
+    def test_fork_grandchildren_have_correct_root(self):
+        fork = self.project.fork_node(auth=self.auth)
+        fork_child = NodeFactory(parent=fork)
+        fork_grandchild = NodeFactory(parent=fork_child)
+        assert_equal(fork_grandchild.root._id, fork._id)
+
+    def test_template_project_has_own_root(self):
+        new_project = self.project.use_as_template(auth=self.auth)
+        assert_equal(new_project.root._id, new_project._id)
+
+    def test_template_project_child_has_correct_root(self):
+        new_project = self.project.use_as_template(auth=self.auth)
+        new_project_child = NodeFactory(parent=new_project)
+        assert_equal(new_project_child.root._id, new_project._id)
+
+    def test_template_project_grandchild_has_correct_root(self):
+        new_project = self.project.use_as_template(auth=self.auth)
+        new_project_child = NodeFactory(parent=new_project)
+        new_project_grandchild = NodeFactory(parent=new_project_child)
+        assert_equal(new_project_grandchild.root._id, new_project._id)
+
+    def test_node_find_returns_correct_nodes(self):
+        # Build up a family of nodes
+        child_node_one = NodeFactory(parent=self.project)
+        child_node_two = NodeFactory(parent=self.project)
+        NodeFactory(parent=child_node_one)
+        NodeFactory(parent=child_node_two)
+        # Create a rogue node that's not related at all
+        NodeFactory()
+
+        family_ids = [self.project._id] + [r._id for r in self.project.get_descendants_recursive()]
+        family_nodes = Node.find(Q('root', 'eq', self.project._id))
+        number_of_nodes = family_nodes.count()
+
+        assert_equal(number_of_nodes, 5)
+        found_ids = []
+        for node in family_nodes:
+            assert_in(node._id, family_ids)
+            found_ids.append(node._id)
+        for node_id in family_ids:
+            assert_in(node_id, found_ids)
+
+    def test_get_descendants_recursive_returns_in_depth_order(self):
+        # Build up a family of nodes
+        node_structure = [5, 2, [1, 2], 3, [10, 2, 1], 2, [4, [2, 2], 6]]
```

Comment:
Again, this complex structure makes this test slow and hard to debug. See previous comments on simplifying.